docs/language-reference/07-declarations.md

```diff
@@ -0,0 +1,771 @@
+
+An initializer declaration may only appear in the body of an interface or a structure type.
+An initializer defines a method for initializing an instance of the enclosing type.
+
+> Note: A C++ programmer might think of an initializer declaration as similar to a C++ _constructor_.
+
+An initializer has a parameter list and body just like a function declaration.
+An initializer must not include a result type clause; the result type of an initializer is always the enclosing type.
+
+An initializer is invoked by calling the enclosing type as if it were a function.
+E.g., in the example above, the initializer in `MyVector` can be invoked as `MyVector(1.0f)`.
+
+
+An initializer has access to an implicit `this` variable that is the instance being initialized; an initializer must not be marked `static`.
+The `this` variable of an initializer is always mutable; an initializer need not, and must not, be marked `[mutating]`.
+
+> Note: Slang currently does not enforce that a type with an initializer can only be initialized using its initializers.
+> It is possible for user code to declare a variable of type `MyVector` above, and explicitly write to the `x` and `y` fields to initialize it.
+> A future version of the language may close up this loophole.
+
+> Note: Slang does not provide any equivalent to C++ _destructors_ which run automatically when an instance goes out of scope.
+
+Subscripts
+----------
+
+A subscript declaration is introduced with the `__subscript` keyword:
+
+```hlsl
+struct MyVector
+{
+    ...
+
+    __subscript(int index) -> float
+    {
+        get { return index == 0 ? x : y; }
+    }
+}
+```
+
+> Note: subscript declarations are a non-finalized and unstable feature, as indicated by the double-underscore (`__`) prefix on the keyword.
+> Arbitrary changes to the syntax and semantics of subscript declarations may be introduced in future versions of Slang.
+
+A subscript declaration introduces a way for a user-defined type to support subscripting with the `[]` braces:
+
+```hlsl
+MyVector v = ...;
+float f = v[0];
+```
+
+A subscript declaration lists one or more parameters inside parantheses, followed by a result type clause starting with `->`.
+The result type clause of a subscript declaration cannot be elided.
+
+The body of a subscript declaration consists of _accessor declarations_.
+Currently only `get` accessor declarations are supported for user code.
+
+A `get` accessor declaration introduces a _getter_ for the subscript.
+The body of a getter is a code block like a function body, and must return the appropriate value for a subcript operation.
+The body of a getter can access the parameters of the enclosing subscript, as a well as an implicit `this` parameter of the type that encloses the accessor.
+The `this` parameter of a getter is immutable; `[mutating]` getters are not currently supported.
+
+Extensions
+----------
+
+An extension declaration is introduced with the `extension` keyword:
+
+```hlsl
+extension MyVector
```

Comment:
You might want to define MyVector in this example - to make clear what is going on here.

Reply:
Fair enough, I had hoped that the example that defined it was close enough that I could call back to it, but maybe it helps to have each section stand on its own.